### PR TITLE
Canvas: Connection properties based on data

### LINKED ISSUE #58384: Canvas: Arrow properties based on data
Similar to geomap marker ability to change size / color based on data, add basic tie ins to properties of the arrow such as size (width, etc.), color, and any other properties that make sense for the initial integration.
```[tasklist]

- [x] Connection selection
- [x] Add color and width properties
- [ ] Update the inline editor with selected connection
- [x] Migration
- [x] Styling for selected connection
```

### PR DESCRIPTION
**Pull request by adela-almasan**
Enables modifying connection property in panel edit mode - we experienced complications with getting the inline editor working due to the overly complex state management causing issues with multiple canvas panels in dashboard view. Note: we still support deleting connections from panel edit mode.

For now we will only support modifying connection properties in panel edit mode as a stop gap as having this functionality in some form is better than delaying for the perfect solution. In the short term we will investigate options to enable modifying connection properties in the dashboard context. Longer term we will want to resolve the connection state management and ideally manage it through selecto to greatly simplify things (see [this experimental branch](https://github.com/grafana/grafana/pull/64360#issuecomment-1518184321)).

Video

https://user-images.githubusercontent.com/22381771/233729717-dee05b42-b155-48b1-8005-6db1186a5d5b.mov

Stress test (doesn't seem to break anything 🤷‍♂️)
Screenshot 2023-04-21 at 1 59 05 PM

Fixes #58384

### CODE REVIEW

File: public/app/features/canvas/element.ts [hunk 1]
```diff
@@ -46,6 +46,8 @@ export interface CanvasConnection {
   target: ConnectionCoordinates;
   targetName?: string;
   path: ConnectionPath;
+  color?: ColorDimensionConfig;
+  size?: number;
```
Comment (ryantxu) [inline public/app/features/canvas/element.ts:50]: size should likely be `ScaleDimensionConfig `

File: public/app/plugins/panel/canvas/CanvasPanel.tsx [hunk 2]
```diff
@@ -130,6 +131,32 @@ export class CanvasPanel extends Component<Props, State> {
           },
         })
       );
+
+      this.subs.add(
+        this.scene.connections.selection.subscribe({
+          next: (v) => {
+            if (!this.context.instanceState) {
```
Comment (ryantxu) [inline public/app/plugins/panel/canvas/CanvasPanel.tsx:138]: likely can avoid this by moving the initialization later in the setup

File: public/app/plugins/panel/canvas/ConnectionSVG.tsx [hunk 3]
```diff
@@ -26,18 +24,17 @@ export const ConnectionSVG = ({ setSVGRef, setLineRef, scene }: Props) => {
...
 
   // Need to use ref to ensure state is not stale in event handler
   const selectedConnectionRef = useRef(selectedConnection);
   useEffect(() => {
     selectedConnectionRef.current = selectedConnection;
   });
 
-  const [selectedConnectionSource, setSelectedConnectionSource] = useState<ElementState | undefined>(undefined);
-  const selectedConnectionSourceRef = useRef(selectedConnectionSource);
+  const selectedConnectionSourceRef = useRef(selectedConnection?.source);
```
Comment (ryantxu) [inline public/app/plugins/panel/canvas/ConnectionSVG.tsx:35]: is this still necessary?

Comment (adela-almasan) [reply public/app/plugins/panel/canvas/ConnectionSVG.tsx]: no, it's removed now!

**Comment by nmarrs** [comment]
Created an experimental branch looking into using moveable to manage connection state - https://github.com/grafana/grafana/tree/canvas-experimental-refactor

see this [moveable issue](https://github.com/daybrush/moveable/issues/848)

**Comment by adela-almasan** [comment]
Tested locally and it looks good in panel edit mode! 🎉


In dashboard view mode I found 2 issues while testing:
1. First one seems related to [this](https://github.com/grafana/grafana/issues/65298) bug - after entering panel edit mode and then getting back to dashboard view mode, the previous panel connection selection doesn't get cleared. Pressing delete, will delete both selections and will trigger issue no. 2. One option would be to block any type of changes to the connections in dashboard mode for now, but not sure how appreciated that would be..

https://user-images.githubusercontent.com/88068998/234333437-fabb1469-fea0-4660-95c2-ad54fabfe405.mov




2. On a random key press the entire dashboard scene seems to get "selected". You can't clear the selection unless you refresh the page or enter panel edit mode. It happens on `main` too. I'll open an issue for this one.

selection


As they don't seem related to this feature, I think this should be good to go!

**Comment by nmarrs** [comment]
> UPDATE - This seems to happen for all dashboards, so it's not related to Canvas. But since we're deleting, it's more obvious.. Not sure where to redirect the issue thought.

I think if it's impacting all dashboards we could try to send it over to either the dashboards squad or potentially user essentials (now called frontend platform)

**Review by nmarrs — APPROVED** [review]
🚀

**Comment by R1d3rBG** [comment]
Hi, are this could be enable in Azure Grafana?

**Pull request merged**